wgsl/index.bs

```diff
@@ -942,9 +1023,281 @@ a result value, such as the non-result-value effects of its subexpressions.
+</table>
+
+### Overload Resolution ### {#overload-resolution-section}
+
+When more than one [=type rule applies to a syntactic phrase=], a tie-breaking procedure is used
+to determine which one should take effect.
+This procedure is called <dfn noexport>overload resolution</dfn>,
+and assumes type checking has already succeeded in finding static types for [=subexpressions=].
+
+Consider a [=syntactic phrase=] |P|, and all [=type rule applies to a syntactic phrase|type rules that apply=] to |P|.
+The overload resolution algorithm calls these type rules <dfn noexport>overload candidates</dfn>.
+For each candidates:
+* Its preconditions have been met.
+* Its [=type rule conclusion|conclusion=] has:
+    * A syntactic form matching a valid parse of |P|, and
+    * A [=type assertion=] corresponding to each [=top-level expression=] in |P|.
+
+Overload resolution for |P| proceeds as follows, with the goal of finding a single most preferable [=overload candidate=]:
+
+1. For each candidate |C|, enumerate conversion ranks for subexpressions in the syntactic phrase.
+    The candidate's preconditions have been met, and so for the |i|'th subexpression in the |P|:
+    * Its static type has computed.
+    * There is a [=feasible automatic conversion=] from the expression's static type to the type required by the corresponding type assertion in the preconditions.
+        Let |C|.|R|(i) be the [=ConversionRank=] of that conversion.
+
+1. Rank candidates: Given two overload candidates |C1| and |C2|, |C1| is <dfn>better</dfn> than |C2| if:
+    * For each expression position |i| in |P|, |C1|.|R|(i) &le; |C2|.|R|(i).
+         * That is, each expression conversion required to apply |C1| to |P| is at least as good as the conversion required to apply |C2| to |P|.
+    * There is at least one expression position |i| where |C1|.|R|(i) &lt; |C2|.|R|(i).
+         * That is, there is at least one expression conversion required to apply |C1| that is preferable to a conversion required to apply |C2|.
+
+1. If there is a single candidate |C| which is [=better=] than all the others, then overload resolution succeeds, yielding the candidate type rule |C|.
+      Otherwise, overload resolution fails.
+
+TODO: Examples
+
+## Types for Creation-Time Constants ## {#types-for-creation-time-constants}
+
+Certain expressions are evaluated at [=shader module creation|shader-creation time=],
+and with a numeric range and precision that may be larger than directly implemented by the GPU.
+
+[SHORTNAME] defines two <dfn>abstract numeric types</dfn> for these evaluations:
+* The <dfn noexport>AbstractInt</dfn> type is the set of integers |i|, with -2<sup>63</sup> &leq; |i| &lt; 2<sup>63</sup>.
+* The <dfn noexport>AbstractFloat</dfn> type is the set of finite floating point numbers representable
+    in the [[!IEEE-754|IEEE-754]] binary64 (double precision) format.
+
+An evaluation of an expression in one of these types must not overflow or be a division by zero.
```

Comment:
Nit: we don't currently have division over abstract numerics. I agree we'll want something in the future, but it seems odd to call out only one case. Maybe say expressions on these types must not overflow or produced undefined results?